### PR TITLE
Improve performance of BitSet.iterator

### PR DESCRIPTION
**Pull request by linasm**
Improves performance of `BitSet.iterator` by utilising `Long.numberOfTrailingZeros` (instead of iterating through all integers in range and checking their presence in the `BitSet`). Partially solves https://github.com/scala/bug/issues/11418.

Benchmarks (`spacing` is the number of `0`s between `1`s, so `spacing = 0` is `11111...`, `spacing = 1` is `101010...`, etc.).
```
Original:
[info] Benchmark                           (spacing)  Mode  Cnt     Score     Error  Units
[info] BitSetIteratorBenchmark.iterateAll          0  avgt    6  7283.606 ±  74.967  ns/op
[info] BitSetIteratorBenchmark.iterateAll          1  avgt    6  5933.497 ±  28.653  ns/op
[info] BitSetIteratorBenchmark.iterateAll          3  avgt    6  3371.763 ±  21.047  ns/op
[info] BitSetIteratorBenchmark.iterateAll         15  avgt    6  2191.009 ±  10.818  ns/op
[info] BitSetIteratorBenchmark.iterateAll         63  avgt    6  1664.190 ±  11.116  ns/op
[info] BitSetIteratorBenchmark.iterateAll        255  avgt    6  1614.554 ±  96.813  ns/op

Modified:
[info] Benchmark                           (spacing)  Mode  Cnt     Score    Error  Units
[info] BitSetIteratorBenchmark.iterateAll          0  avgt    6  3526.702 ± 36.208  ns/op
[info] BitSetIteratorBenchmark.iterateAll          1  avgt    6  2953.840 ± 29.800  ns/op
[info] BitSetIteratorBenchmark.iterateAll          3  avgt    6  1442.062 ± 11.518  ns/op
[info] BitSetIteratorBenchmark.iterateAll         15  avgt    6   395.849 ±  2.110  ns/op
[info] BitSetIteratorBenchmark.iterateAll         63  avgt    6   127.068 ±  0.477  ns/op
[info] BitSetIteratorBenchmark.iterateAll        255  avgt    6    53.174 ±  2.066  ns/op
```

Benchmark setup:
```
[info] # JMH version: 1.20
[info] # VM version: JDK 11.0.4, VM 11.0.4+11
[info] # VM invoker: /Library/Java/JavaVirtualMachines/adoptopenjdk-11.jdk/Contents/Home/bin/java
[info] # VM options: <none>
```

Benchmark code:
```
package scala.collection.mutable

import java.util.concurrent.TimeUnit

import org.openjdk.jmh.annotations._
import org.openjdk.jmh.infra._

import scala.collection.mutable

@BenchmarkMode(Array(Mode.AverageTime))
@Fork(1)
@Threads(1)
@Warmup(iterations = 6)
@Measurement(iterations = 6)
@OutputTimeUnit(TimeUnit.NANOSECONDS)
@State(Scope.Benchmark)
class BitSetIteratorBenchmark {

  @Param(Array("0", "1", "3", "15", "63", "255"))
  var spacing: Int = _

  var bs: mutable.BitSet = _

  @Setup(Level.Iteration) def initializeRange(): Unit = {
    bs = mutable.BitSet(0 until 1000 by (spacing + 1): _*)
  }

  @Benchmark def iterateAll(): Unit = {
    var sum = 0
    val it = bs.iterator
    while (it.hasNext) sum += it.next
  }

}
```

### CODE REVIEW

File: src/library/scala/collection/BitSet.scala [hunk 1]
```diff
@@ -108,15 +108,22 @@ trait BitSetOps[+C <: BitSet with BitSetOps[C]]
...
+      }
+      currentWord != 0
+    }
+    def next(): Int = {
+      if (hasNext) {
+        val bitPos = java.lang.Long.numberOfTrailingZeros(currentWord)
+        currentWord &= currentWord - 1
+        (currentPos << LogWL) + bitPos
+      } else Iterator.empty.next()
     }
```
Comment (viktorklang) [inline src/library/scala/collection/BitSet.scala:127]: Cool stuff @linasm!

Does a minor tweaked solution like the following offer any benefits performance-wise?

WARNING: FOLLOWING CODE HAS NEVER BEEN COMPILED

```scala
def iteratorFrom(start: Int): Iterator[Int] = new AbstractIterator[Int] {
    private[this] final var currentPos = math.max(start >> LogWL, 0)
    private[this] final var currentWord = if (start > 0) word(currentPos) & (-1L << (start & (WordLength - 1))) else word(0)
    @tailrec private[this] final def nextWord(): Int = {
      if (currentWord == 0) {
        if (currentPos + 1 < nwords) {
          currentPos += 1
          currentWord = word(currentPos)
          nextWord()
        } else 0
      } else currentWord
    }
    final override def hasNext: Boolean = nextWord() != 0
    final override def next(): Int = {
      val word = nextWord()
      if (word != 0) {
        currentWord &= word - 1
        (currentPos << LogWL) + java.lang.Long.numberOfTrailingZeros(word)
      }
    }
  }
```

Comment (linasm) [reply src/library/scala/collection/BitSet.scala:127]: @viktorklang thanks for your suggestions! I've tried benchmarking the suggested implementation, and it really gives a nice further improvement:
```
[info] Benchmark                           (spacing)  Mode  Cnt     Score    Error  Units
[info] BitSetIteratorBenchmark.iterateAll          0  avgt    6  3591.702 ± 30.049  ns/op
[info] BitSetIteratorBenchmark.iterateAll          1  avgt    6  1782.208 ± 19.453  ns/op
[info] BitSetIteratorBenchmark.iterateAll          3  avgt    6   896.987 ± 11.696  ns/op
[info] BitSetIteratorBenchmark.iterateAll         15  avgt    6   240.771 ±  2.595  ns/op
[info] BitSetIteratorBenchmark.iterateAll         63  avgt    6    77.982 ±  0.642  ns/op
[info] BitSetIteratorBenchmark.iterateAll        255  avgt    6    36.464 ±  0.428  ns/op
```
However, you helped me realize what can be improved in my implementation, and I was able to get basically the same (within ±1% margin) improvements, with less (and arguably simpler) code (updated the PR):
```
[info] Benchmark                           (spacing)  Mode  Cnt     Score    Error  Units
[info] BitSetIteratorBenchmark.iterateAll          0  avgt    6  3560.275 ± 65.491  ns/op
[info] BitSetIteratorBenchmark.iterateAll          1  avgt    6  1792.763 ± 13.405  ns/op
[info] BitSetIteratorBenchmark.iterateAll          3  avgt    6   908.467 ±  7.259  ns/op
[info] BitSetIteratorBenchmark.iterateAll         15  avgt    6   242.321 ±  3.470  ns/op
[info] BitSetIteratorBenchmark.iterateAll         63  avgt    6    78.177 ±  0.852  ns/op
[info] BitSetIteratorBenchmark.iterateAll        255  avgt    6    36.429 ±  0.200  ns/op
```
Please let me know what you think, thanks.

File: src/library/scala/collection/BitSet.scala [hunk 2]
```diff
@@ -108,15 +108,23 @@ trait BitSetOps[+C <: BitSet with BitSetOps[C]]
...
-    private[this] var current = start
-    private[this] val end = nwords * WordLength
-    def hasNext: Boolean = {
-      while (current != end && !self.contains(current)) current += 1
-      current != end
+    private[this] var currentPos = if (start > 0) start >> LogWL else 0
+    private[this] var currentWord = if (start > 0) word(currentPos) & (-1L << (start & (WordLength - 1))) else word(0)
+    final override def hasNext: Boolean = {
+      while (currentWord == 0) {
+        if (currentPos + 1 >= nwords) return false
```
Comment (viktorklang) [inline src/library/scala/collection/BitSet.scala:115]: @linasm I'm not a fan of `return` in Scala as it breaks last-expr-is-the-result assumptions.

Comment (linasm) [reply src/library/scala/collection/BitSet.scala:115]: @viktorklang me neither, but I feel similar about `tailrec` method that does side effects :)
Before submitting this change, I saw `return` from `while` all over `BitSet` implementation:
https://github.com/scala/scala/blob/4ec48ff29755fc01fdde880dbb4548085420c601/src/library/scala/collection/BitSet.scala#L148-L151
https://github.com/scala/scala/blob/4ec48ff29755fc01fdde880dbb4548085420c601/src/library/scala/collection/BitSet.scala#L160-L163
https://github.com/scala/scala/blob/4ec48ff29755fc01fdde880dbb4548085420c601/src/library/scala/collection/mutable/BitSet.scala#L246-L247
https://github.com/scala/scala/blob/4ec48ff29755fc01fdde880dbb4548085420c601/src/library/scala/collection/mutable/BitSet.scala#L253-L254
Not implying that this legitimizes the pattern or something, just would appreciate a second opinion from someone who maintains this part of the code base.

Comment (viktorklang) [reply src/library/scala/collection/BitSet.scala:115]: @linasm I think "prior art" is a valid argument. :)

**Comment by szeiger** [comment]
Can you add the benchmark code under `test/benchmarks`?

File: src/library/scala/collection/BitSet.scala (hunk 2 above)
Comment (diesalbla) [reply src/library/scala/collection/BitSet.scala:115]: IMHO, while "prior art" is a fair enough reason, there is no reason not to "clean" it along the way, unless it defeats performance of course. I think the following should work (but please do test first).

```Scala
      while (currentWord == 0 && currentPos + 1 < nwords) {
        currentPos += 1
        currentWord = word(currentPos)
      }
      currentWord != 0 || currentPos + 1 < nwords
```

File: src/library/scala/collection/BitSet.scala [hunk 3]
```diff
@@ -108,15 +108,23 @@ trait BitSetOps[+C <: BitSet with BitSetOps[C]]
...
+        currentPos += 1
+        currentWord = word(currentPos)
+      }
+      true
+    }
+    final override def next(): Int = {
+      if (hasNext) {
+        val bitPos = java.lang.Long.numberOfTrailingZeros(currentWord)
+        currentWord &= currentWord - 1
+        (currentPos << LogWL) + bitPos
```
Comment (szeiger) [inline src/library/scala/collection/BitSet.scala]: Nice idea. I was wondering how you get away with only storing the current word but no index into it until I saw this.

File: src/library/scala/collection/BitSet.scala [hunk 4]
```diff
@@ -108,15 +108,23 @@ trait BitSetOps[+C <: BitSet with BitSetOps[C]]
...
+        if (currentPos + 1 >= nwords) return false
+        currentPos += 1
+        currentWord = word(currentPos)
+      }
+      true
+    }
+    final override def next(): Int = {
+      if (hasNext) {
+        val bitPos = java.lang.Long.numberOfTrailingZeros(currentWord)
+        currentWord &= currentWord - 1
```
Comment (szeiger) [inline src/library/scala/collection/BitSet.scala:124]: Nice idea. I was wondering how you get away with only storing the current word but no index into it until I saw this.

File: test/junit/scala/collection/mutable/BitSetTest.scala [hunk 5]
```diff
@@ -153,6 +153,20 @@ class BitSetTest {
   @Test def buildFromRange(): Unit = {
     import scala.util.chaining._
     assert((1 to 1000).to(BitSet) == BitSet().tap(bs => (1 to 1000).foreach(bs.addOne)))
+  }
 
+  @Test def iteratorFrom(): Unit = {
+    val a = (0 to 127).to(BitSet)
```
Comment (szeiger) [inline test/junit/scala/collection/mutable/BitSetTest.scala:159]: Almost all tests are based on `a`, the only exception being the one with `BitSet(0)`. `a` is quite regular with 2 full words. Can we have some tests with holes in the data or data that does not begin and end on a full word? Maybe use a ScalaCheck test instead of manually coming up with corner cases?

**Comment by linasm** [comment]
> Can you add the benchmark code under `test/benchmarks`?

Done.

File: test/junit/scala/collection/mutable/BitSetTest.scala (hunk 5 above)
Comment (linasm) [reply test/junit/scala/collection/mutable/BitSetTest.scala:159]: Agreed, I've added ScalaCheck test.

File: src/library/scala/collection/BitSet.scala (hunk 2 above)
Comment (linasm) [reply src/library/scala/collection/BitSet.scala:115]: @diesalbla your implementation seems to be correct, but it is slightly slower in most cases:
```
[info] Benchmark                           (spacing)  Mode  Cnt     Score    Error  Units
[info] BitSetIteratorBenchmark.iterateAll          0  avgt    6  3561.769 ± 79.572  ns/op
[info] BitSetIteratorBenchmark.iterateAll          1  avgt    6  1786.156 ± 11.039  ns/op
[info] BitSetIteratorBenchmark.iterateAll          3  avgt    6   934.819 ± 13.076  ns/op
[info] BitSetIteratorBenchmark.iterateAll         15  avgt    6   253.551 ±  5.492  ns/op
[info] BitSetIteratorBenchmark.iterateAll         63  avgt    6    83.190 ±  0.820  ns/op
[info] BitSetIteratorBenchmark.iterateAll        255  avgt    6    38.085 ±  3.758  ns/op
```
`hasNext` is normally invoked twice for each advancement of the iterator (once directly from client code, and once from `next()`), and in most invocations it does not enter the `while` loop. I've optimized my code under this assumption, making sure that just one comparison is done in those cases.

**Review by szeiger — APPROVED** [review]
